packages/uniforms/src/BaseForm.js

```diff
@@ -238,10 +247,16 @@ export default class BaseForm extends Component {
             event.stopPropagation();
         }
 
-        return Promise.resolve(
-            this.props.onSubmit &&
-            this.props.onSubmit(this.getModel('submit'))
-        ).then(
+        const result = this.props.onSubmit && this.props.onSubmit(this.getModel('submit'));
+        let submitting = Promise.resolve(result);
+
+        // Do not change the `submitting` state if onSubmit is not async so we don't cause an unnecessary re-render
+        if (result && isFunction(result.then)) {
+            this.setState({submitting: true});
+            submitting = submitting.finally(() => this.setState({submitting: false}));
+        }
```

Comment:
Now I'm thinking about going even one step further: we can get rid of an additional `Promise` - in the `if` use the `result` directly and use `Promise.resolve` only within the `else` clause.

Reply:
I feel like we would want to keep the `Form#onSubmit` return type consistent, and always return a Promise (it could almost be an `async` function). Otherwise it would essentially have two versions - sync and async - in which case I would want to support catching errors again to support the sync case fully.

Reply:
My proposition won't change it:

```diff
         const result = this.props.onSubmit && this.props.onSubmit(this.getModel('submit'));
-        let submitting = Promise.resolve(result);
+        let submitting;
 
         // Do not change the `submitting` state if onSubmit is not async so we don't cause an unnecessary re-render
         if (result && isFunction(result.then)) {
             this.setState({submitting: true});
-            submitting = submitting.finally(() => this.setState({submitting: false}));
+            submitting = result.finally(() => this.setState({submitting: false}));
+        } else {
+            submitting = Promise.resolve(result);
         }
```

Reply:
I see. I'll use that.